src/compiler/checker.ts

```diff
@@ -8050,6 +8110,10 @@ namespace ts {
                 }
             }
 
+            if (compilerOptions.noImplicitThis && isFunctionLike(container)) {

```

Comment:
we should not need the `isFunctionLike` here. it should be an error to use `this` in the global scope with `noImplicitThis`